### PR TITLE
build canonical url when canonical ref is relative

### PR DESCRIPTION
**Pull request by michaelshobbs**
https://app.asana.com/0/172614131592670/1156315720563139

problem example:
```
<link rel="canonical" href="/image-feature/houston-texas-on-the-gulf-coast" />
```

### CODE REVIEW

File: lib/analyze.js [hunk 1]
```diff
@@ -147,13 +147,20 @@ function gatherMetaData(url, body) {
 
       return foundValue
     }
-
+    const extractCanonical = function() {
+      const canonical = getSimpleValue('link[rel="canonical"]', 'href')
+      if (canonical && canonical.startsWith('/')) {
+        const parsedUrl = Url.parse(url)
+        return `${parsedUrl.protocol}//${parsedUrl.host}${canonical}`
```
Comment (meeq) [inline lib/analyze.js]: Will this handle URLs with a port number?

See: https://www.grc.com/port_81.htm
> Sometimes an "off the beaten path" web server will be established to listen not to port 80, but to an alternate related port such as 81, 82, 8080, or 8090.

Comment (anoakie) [reply lib/analyze.js]: It does.  Host is either hostname or hostname:port

**Comment by michaelshobbs** [comment]
> Should we be defensive for canonical urls?
> We should block javascript: and other non-safe patterns (only allow urls starting with https?:// or /)

@anoakie what would that look like in the context of the problem here?

File: lib/analyze.js [hunk 2]
```diff
@@ -147,13 +147,20 @@ function gatherMetaData(url, body) {
 
       return foundValue
     }
-
+    const extractCanonical = function() {
+      const canonical = getSimpleValue('link[rel="canonical"]', 'href')
```
Comment (anoakie) [inline lib/analyze.js]:
```
if (canonical && !regexp.MatchString(`^(https?://|/)`, canonical)) {
    // error and return
}
```

File: lib/analyze.js [hunk 3]
```diff
@@ -147,13 +147,20 @@ function gatherMetaData(url, body) {
...
+      if (canonical && canonical.startsWith('/')) {
+        const parsedUrl = Url.parse(url)
```
Comment (anoakie) [inline lib/analyze.js]:
```
  const parsedUrl, err = Url.parse(url)
  if (err) {
    // error and return
  }
```

**Comment by anoakie** [comment]
> @anoakie what would that look like in the context of the problem here?

Added examples.  Since `canonical` is a free-form string, we should try to be defensive.

**Comment by michaelshobbs** [comment]
closing in favor of https://github.com/mix/scraper/pull/56

**Pull request closed without merging**